modules/core/src/vaapi.cpp

```diff
@@ -0,0 +1,326 @@
+
+        if (((void*)clGetDeviceIDsFromVA_APIMediaAdapterINTEL == NULL) ||
+            ((void*)clCreateFromVA_APIMediaSurfaceINTEL == NULL) ||
+            ((void*)clEnqueueAcquireVA_APIMediaSurfacesINTEL == NULL) ||
+            ((void*)clEnqueueReleaseVA_APIMediaSurfacesINTEL == NULL))
+        {
+            continue;
+        }
+
+        // Query device list
+
+        cl_uint numDevices = 0;
+
+        status = clGetDeviceIDsFromVA_APIMediaAdapterINTEL(platforms[i], CL_VA_API_DISPLAY_INTEL, display,
+                                                           CL_PREFERRED_DEVICES_FOR_VA_API_INTEL, 0, NULL, &numDevices);
+        if ((status != CL_SUCCESS) || !(numDevices > 0))
+            continue;
+        numDevices = 1; // initializeContextFromHandle() expects only 1 device
+        status = clGetDeviceIDsFromVA_APIMediaAdapterINTEL(platforms[i], CL_VA_API_DISPLAY_INTEL, display,
+                                                           CL_PREFERRED_DEVICES_FOR_VA_API_INTEL, numDevices, &device, NULL);
+        if (status != CL_SUCCESS)
+            continue;
+
+        // Creating CL-VA media sharing OpenCL context
+
+        cl_context_properties props[] = {
+            CL_CONTEXT_VA_API_DISPLAY_INTEL, (cl_context_properties) display,
+            CL_CONTEXT_INTEROP_USER_SYNC, CL_FALSE, // no explicit sync required
+            0
+        };
+
+        context = clCreateContext(props, numDevices, &device, NULL, NULL, &status);
+        if (status != CL_SUCCESS)
+        {
+            clReleaseDevice(device);
+        }
+        else
+        {
+            found = i;
+            break;
+        }
+    }
+
+    if (found < 0)
+        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't create context for VA-API interop");
+
+    Context& ctx = Context::getDefault(false);
+    initializeContextFromHandle(ctx, platforms[found], context, device);
+    return ctx;
+#endif
+}
+
+#if defined(HAVE_VAAPI) && defined(HAVE_OPENCL)
+static bool ocl_convert_nv12_to_bgr(cl_mem clImageY, cl_mem clImageUV, cl_mem clBuffer, int step, int cols, int rows)
+{
+    ocl::Kernel k;
+    k.create("YUV2BGR_NV12_8u", cv::ocl::core::cvtclr_dx_oclsrc, "");
+    if (k.empty())
+        return false;
+
+    k.args(clImageY, clImageUV, clBuffer, step, cols, rows);
+
+    size_t globalsize[] = { cols, rows };
+    return k.run(2, globalsize, 0, false);
+}
+
+static bool ocl_convert_bgr_to_nv12(cl_mem clBuffer, int step, int cols, int rows, cl_mem clImageY, cl_mem clImageUV)
+{
+    ocl::Kernel k;
+    k.create("BGR2YUV_NV12_8u", cv::ocl::core::cvtclr_dx_oclsrc, "");
+    if (k.empty())
+        return false;
+
+    k.args(clBuffer, step, cols, rows, clImageY, clImageUV);
+
+    size_t globalsize[] = { cols, rows };
+    return k.run(2, globalsize, 0, false);
+}
+#endif // HAVE_VAAPI && HAVE_OPENCL
+
+} // namespace cv::vaapi::ocl
+
+void convertToVASurface(InputArray src, VASurfaceID surface, Size size)
```

Comment:
I believe that just check is not enough for this function. What user has to do for platform where VA-CL interoperability does not work? It looks like he must to write ifelse constructions for this 2 cases. I believe It would be better to have some fallback code inside convertToVASurface for case when VA-CL interoperability is not supported by platform.